scrapers_library/data_portals/ckan/scrape_ckan_data_portals.py

```diff
@@ -0,0 +1,284 @@
+from itertools import chain
+import json
+import sys
+from typing import Any, Callable, Optional
+
+from from_root import from_root
+import pandas as pd
+from tqdm import tqdm
+
+p = from_root("CONTRIBUTING.md").parent
+sys.path.insert(1, str(p))
+
+from scrapers_library.data_portals.ckan.ckan_scraper_toolkit import (
+    ckan_package_search,
+    ckan_group_package_show,
+    ckan_collection_search,
+    ckan_package_search_from_organization,
+    Package,
+)
+from search_terms import package_search, group_search, organization_search
+
+
+def perform_search(
+    search_func: Callable,
+    search_terms: list[dict[str, Any]],
+    results: list[dict[str, Any]],
+):
+    """Executes a search function with the given search terms.
+
+    :param search_func: The search function to execute.
+    :param search_terms: The list of urls and search terms.
+    :param results: The list of results.
+    :return: Updated list of results.
+    """
+    key = list(search_terms[0].keys())[1]
+    for search in tqdm(search_terms):
+        results += [search_func(search["url"], item) for item in search[key]]
+
+    return results
+
+
+def get_collection_child_packages(
+    results: list[dict[str, Any]]
+) -> list[dict[str, Any]]:
+    """Retrieves the child packages of each collection.
+
+    :param results: List of results.
+    :return: List of results containing child packages.
+    """
+    new_list = []
+
+    for result in tqdm(results):
+        if "extras" in result.keys():
+            collections = [
+                ckan_collection_search(
+                    base_url="https://catalog.data.gov/dataset/",
+                    collection_id=result["id"],
+                )
+                for extra in result["extras"]
+                if extra["key"] == "collection_metadata"
+                and extra["value"] == "true"
+                and not result["resources"]
+            ]
+
+            if collections:
+                new_list += collections[0]
+                continue
+
+        new_list.append(result)
+
+    return new_list
+
+
+def filter_result(result: dict[str, Any] | Package):
+    """Filters the result based on the defined criteria.
+
+    :param result: The result to filter.
+    :return: True if the result should be included, False otherwise.
+    """
+    if isinstance(result, Package) or "extras" not in result.keys():
+        return True
+
+    for extra in result["extras"]:
+        # Remove parent packages with no resources
+        if (
+            extra["key"] == "collection_metadata"
+            and extra["value"] == "true"
+            and not result["resources"]
+        ):
+            return False
+        # Remove non-public packages
+        elif extra["key"] == "accessLevel" and extra["value"] == "non-public":
+            return False
+
+    # Remove packages with no data or landing page
+    if len(result["resources"]) == 0:
+        landing_page = next(
```

Comment:
Let's extract some of this logic! This is a next, a for loop, two key-value accesses, and a boolean conditional condensed in a very small amount of real estate!